### PR TITLE
UI: View Controls in Standard Panels

### PR DESCRIPTION
**Pull request by alex40724**
Feature Wiki Request:
https://docu.ilias.de/goto_docu_wiki_wpage_5595_1357.html

Rules:
https://github.com/ILIAS-eLearning/ILIAS/pull/2230/files#diff-e6c01f5f06c4826ee5016bd001f1725a

Interfaces Changes:
https://github.com/ILIAS-eLearning/ILIAS/pull/2230/files#diff-b67dde8100f69a0e226202e796a9b002
https://github.com/ILIAS-eLearning/ILIAS/pull/2230/files#diff-43e7e226dfc489d581edcf75c86205dc

Bildschirmfoto 2019-10-08 um 11 56 35

### CODE REVIEW

File: src/UI/Component/Panel/Factory.php [hunk 1]
```diff
@@ -19,8 +19,10 @@ interface Factory
      *   purpose: >
      *      Standard Panels are used in the center content section to group content.
      *   composition: >
-     *      Standard Panels consist of a title and a content section. The
-     *      structure of this content might be varying from Standard
+     *      Standard Panels consist of a content and a top section.
+     *      The top section may contain a title and a sortation view control.
```
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: Sortation View Control (capitalize UI Components)

File: src/UI/Component/Panel/Factory.php [hunk 2]
```diff
@@ -19,8 +19,10 @@ interface Factory
...
+     *      The content section may contain a Pagination view controller or a Section view controller.
```
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: ... Pagination View Control or a Section View Control.

File: src/UI/Component/Panel/Factory.php [hunk 3]
```diff
@@ -30,6 +32,13 @@ interface Factory
      *   usage:
      *      1: In Forms Standard Panels MUST be used to group different sections into Form Parts.
      *      2: Standard Panels SHOULD be used in the center content as primary Container for grouping content of varying content.
+     *
+     *   composition:
+     *      1: Standard Panels MAY contain a Section view control to change the current presentation of the content.
```
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: Section View Control

File: src/UI/Component/Panel/Factory.php [hunk 4]
```diff
@@ -30,6 +32,13 @@ interface Factory
...
+     *      2: Standard Panels MAY contain a Pagination view control to display data in chunks.
```
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: Pagination View Control

File: src/UI/Component/Panel/Factory.php [hunk 5]
```diff
@@ -30,6 +32,13 @@ interface Factory
...
+     *      3: Standard Panels MUST NOT contain more than one UI ViewControl in the main content.
```
Comment (Amstutz) [inline src/UI/Component/Panel/Factory.php]: View Control (add space)

File: src/UI/Component/Panel/PanelViewControls.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+/* Copyright (c) 1998-2019 ILIAS open source, Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Panel;
+
+/**
+ * This describes a Standard Panel.
```
Comment (Amstutz) [inline src/UI/Component/Panel/PanelViewControls.php]: Does it? ;-)

File: src/UI/Implementation/Component/Panel/PanelViewControls.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * @var array
+     */
+    protected $view_controls;
+
+    /**
+     * @param array $view_controls
+     * @return \ILIAS\UI\Component\Component
+     */
+    public function withViewControls(array $view_controls) : \ILIAS\UI\Component\Component
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/PanelViewControls.php]: add `@inheritdocs`

File: src/UI/Implementation/Component/Panel/PanelViewControls.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        $clone = clone $this;
+        $clone->view_controls = $view_controls;
+        return $clone;
+    }
+    /**
+     * @return array|null
+     */
+    public function getViewControls(): ?array
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Panel/PanelViewControls.php]: add `@inheritdocs`

**Review by Amstutz — COMMENTED** [review]
Just some tiny formal issues/typos that popped up during reading through the PR. For the ease of readability I did comment them inline. The formal review with the usual template follows shortly.

**Comment by klees** [comment]
Hi @alex40724!

Thanks a lot for this contribution to the UI-framework and especially for the refactoring while moving forward.

This looks LGTM in general and should in our opinion be uncontroversial.

Still, please answer the following questions:
- [ ] Why are some of the view controls put in a top section and some in the content-section?
- [ ] What is the top section, anyway?
- [ ] Why can the content-section only contain a Pagination View Control or a Section View Control, not both?

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [ ] Be less specific regarding the concrete view controls that are allowed. If we add View Controls, we would need to add them here. The interface on the otherhand tells otherwise and allows all view controls. This slipped our attention when the View Controls where introduced for the Secondary Panel and imo should be adjusted there as well.
- [ ] Be less specific about the placement of View Controls in top or content section in the KS-entry. To us this seems to be just arbitrary (or the answers to the questions tell us otherwise =) Just provide the according LESS that puts them there. This will open wiggle space for future harmonization in how View Controls are displayed above $content. This applies to the Secondary Panel as well. Or is this essential?

Please implement the following changes:
- [ ] Rename `PanelViewControls` to `HasViewControls`, move it to the `ViewControls`-namespace and use it in the `Presentation Table` as well. We guess this trait will come in handy in even more locations. Thanks for introducing it.
- [ ] Add some Unit Test making sure, that View Controls can properly added to and rendered by Standard Panels
- [ ] Correct the small formal issues/typos commented inline by @amstutz above.

Best regards!
@klees, @amstutz

**Comment by alex40724** [comment]
Thanks for the quick feedback.

> Why are some of the view controls put in a top section and some in the content-section?
(and all other questions related to the top / content section).

All these sentences / rules related to top / content section are based on the rules for the Secondary Panels:

https://github.com/ILIAS-eLearning/ILIAS/blob/e14c4c5cd7c5abd42507fafbc2d275859d89278d/src/UI/Component/Panel/Factory.php#L147

So my assumption was, that the definition of these sections was alreay clear.

I do not need this and can simply go with "Panels may have view controls." Should this also be changed in the Secondary Panels rules?

**Comment by Amstutz** [comment]
IMO, if it is fine for you to take the change in sec. along, it would be great. If not let me know, I can do.

File: src/UI/Component/Panel/Factory.php (hunk 1 above)
Comment (alex40724) [reply src/UI/Component/Panel/Factory.php]: Done

File: src/UI/Component/Panel/Factory.php (hunk 2 above)
Comment (alex40724) [reply src/UI/Component/Panel/Factory.php]: Removed due to feedback from @klees

File: src/UI/Component/Panel/Factory.php (hunk 3 above)
Comment (alex40724) [reply src/UI/Component/Panel/Factory.php]: Done

File: src/UI/Component/Panel/Factory.php (hunk 4 above)
Comment (alex40724) [reply src/UI/Component/Panel/Factory.php]: Done

File: src/UI/Component/Panel/Factory.php (hunk 5 above)
Comment (alex40724) [reply src/UI/Component/Panel/Factory.php]: REmoved due to feedback from @klees

File: src/UI/Component/Panel/PanelViewControls.php (hunk 6 above)
Comment (alex40724) [reply src/UI/Component/Panel/PanelViewControls.php]: Fixed

File: src/UI/Implementation/Component/Panel/PanelViewControls.php (hunk 7 above)
Comment (alex40724) [reply src/UI/Implementation/Component/Panel/PanelViewControls.php]: Done

File: src/UI/Implementation/Component/Panel/PanelViewControls.php (hunk 8 above)
Comment (alex40724) [reply src/UI/Implementation/Component/Panel/PanelViewControls.php]: Done

**Comment by alex40724** [comment]
I added all requested changes, this should be sufficient for todays JF discussion on the interface.

Still missing for implementation:
- [ ] Changes in Presentation Table
- [ ] Example
- [ ] Unit Tests

**Comment by matthiaskunkel** [comment]
Jour Fixe, 21 OCT 2019 : We highly appreciate this suggestion and accept the PR for trunk. Please feel free to merge.

**Comment by klees** [comment]
@alex40724 You are still working on this, right? Ping me/Timon if you need anything here. Thx!

**Comment by alex40724** [comment]
@klees @Amstutz 

The changes in Presentation Table are added, together with an example and unit tests.

Hope this can be merged now.

**Review by Amstutz — APPROVED** [review]
Thx for the examples, thx for the IMO good set of tests, thx for improving the existing panels and finally thx for implementing this by creating a tread. IMO good work, @klees : merge a soon as ready.

**Pull request merged**